### ISSUE TITLE
Runtime error with list of Gaussian Random Walks

### ISSUE THREAD

**Issue by sachinruk** [pymc-devs/pymc#1838]
I wish to model the growth in sales of different countries where each country is given a seperate GaussianRandomWalk. The model is as follows (note the block I am having trouble with):

```
t_t = theano.shared(t.ravel())
y_t = theano.shared(y.ravel())
time_slot = theano.shared(big_df.time_slot.values)
country_idx = theano.shared(big_df.idx.values)

with pm.Model() as model:
    sigma_alpha = pm.Exponential('sigma_alpha', 1./.1, testval = .1, shape=n_country)
    sigma_beta = pm.Exponential('sigma_beta', 1./.1, testval = .1, shape=n_country)
#####################
# Part I need help with
#####################
    alpha = []
    beta = []
    for i in range(n_countries):
        alpha.append(pm.GaussianRandomWalk('alpha%d'%(i), sd=sigma_alpha[i], shape=(1,time_slots)))
        beta.append(pm.GaussianRandomWalk('beta%d'%(i), sd=sigma_beta[i], shape=(1,time_slots)))
    alpha = T.concatenate(alpha)
    beta = T.concatenate(beta)
########################    
    alpha_r = alpha[country_idx, time_slot]
    beta_r = beta[country_idx, time_slot]
    
    f = alpha_r + beta_r*t_t
    
    sd = pm.Uniform('sd',lower=0,upper=1,testval=T.std(y_t))
    likelihood = pm.Normal('y',mu=f,sd=sd,observed=y_t)
```
I assumed that `T.concatenate` part would create a (n_country, time_slot) matrix. However when I run ADVI on this I get the following error:

```
===============================
===============================
/Users/sachin/.theano/compiledir_Darwin-15.6.0-x86_64-i386-64bit-i386-3.5.2-64/tmp4ufylpps/mod.cpp:7179:6: fatal error: bracket nesting level exceeded maximum of 256
    {Py_XINCREF(py_V505);}
     ^
/Users/sachin/anaconda/include/python3.5m/object.h:836:8: note: expanded from macro 'Py_XINCREF'
    do {                                              \
       ^
/Users/sachin/.theano/compiledir_Darwin-15.6.0-x86_64-i386-64bit-i386-3.5.2-64/tmp4ufylpps/mod.cpp:7179:6: note: use -fbracket-depth=N to increase maximum nesting level
/Users/sachin/anaconda/include/python3.5m/object.h:836:8: note: expanded from macro 'Py_XINCREF'
    do {                                              \
       ^
1 error generated.
```

In case you want my ADVI code it is as follows:
```
# Tensors and RV that will be using mini-batches
minibatch_tensors = [t_t, time_slot, country_idx, y_t]
minibatch_RVs = [likelihood]

# Generator that returns mini-batches in each iteration
def create_minibatch(data):
    rng = np.random.RandomState(0)

    while True:
        # Return random data samples of set size 100 each iteration
        ixs = rng.randint(len(data), size=256)
        yield data[ixs]

minibatches = zip(
    create_minibatch(t.ravel()),
    create_minibatch(big_df.time_slot.values),
    create_minibatch(big_df.idx.values),
    create_minibatch(y.ravel())
)

with model:
    # Run advi_minibatch
    v_params = pm.variational.advi_minibatch(
        n=50000, minibatch_tensors=minibatch_tensors,
        minibatch_RVs=minibatch_RVs, minibatches=minibatches,
        total_size=len(y), learning_rate=1e-2, epsilon=1.0
    )
```
p.s. This did work when I replaced the block that I was having trouble with, with:
```
alpha = pm.GaussianRandomWalk('alpha', sd=sigma_alpha, shape=(n_country,time_slots)
beta = pm.GaussianRandomWalk('beta', sd=sigma_alpha, shape=(n_country,time_slots)
```
but I want a different `sigma_alpha` for each country.

**Comment by AustinRochford** [comment]
What happens with

```
alpha = pm.GaussianRandomWalk('alpha', sd=sigma_alpha, shape=(n_country,time_slots)
beta = pm.GaussianRandomWalk('beta', sd=sigma_alpha, shape=(n_country,time_slots)
```

if you make `sigma_alpha` have shape `(n_country, 1)`?

**Comment by sachinruk** [comment]
It complained about a dimensionality mismatch. I had to change the shape of `sigma_alpha` to `(n_country-1, 1)` for it to work. This to me suggests that the GaussianRandomWalk is connected via the countries, which is not what I intended. 

I did get it to compile and run after I did (note all the transposes `T` and shape of alpha):
```
    sigma_alpha = pm.Exponential('sigma_alpha', 1./.1, testval = .1, shape=(n_country,1))
    sigma_beta = pm.Exponential('sigma_beta', 1./.1, testval = .1, shape=(n_country,1))
    alpha = pm.GaussianRandomWalk('alpha', sd=sigma_alpha.T, shape=(time_slots, n_country))
    beta = pm.GaussianRandomWalk('beta', sd=sigma_beta.T, shape=(time_slots, n_country))
    
    alpha_r = alpha.T[country_idx, time_slot]
    beta_r = beta.T[country_idx, time_slot]
```
**However**, it seems that the posterior on sigma_beta is constant regardless of the country which suggests that there is some funny business going on here.

In conclusion it seems that the concatenate option is the safest option.

**Comment by sachinruk** [comment]
So in the end this is how I solved it. Had to copy some of mvGaussianRandomWalk that is set to be released in pymc3 v3.1 into here (and no I didnt install pymc3 v3.1). The biggest problem previously was that the posterior on sigma_beta was the same regardless of country, and was concetrated on small numbers. This made the betas constant regardless of the random walk behaviour. Now the problem has been rectified.

Firstly define the class:
```
from pymc3.distributions import Continuous

class mvNormalRandomWalk(Continuous):
    def __init__(self, mu=0., taus=1., *args, **kwargs):
        super(mvNormalRandomWalk, self).__init__(*args, **kwargs)
        self.tau = T.nlinalg.diag(taus)
        self.mu = mu
        self.mean = 0.

    def logp(self, x):
        tau = self.tau
        mu = self.mu

        x_im1 = x[:-1]
        x_i = x[1:]

        innov_like = pm.MvNormal.dist(mu=x_im1 + mu, tau=tau).logp(x_i)
        return T.sum(innov_like)
```

```
t = big_df.t.values[:,None]
t_scaler = MinMaxScaler()
t = t_scaler.fit_transform(t)

# Theano variables
t_t = theano.shared(t.ravel())
y_t = theano.shared(y.ravel())
time_slot = theano.shared(big_df.time_slot.values)
country_idx = theano.shared(big_df.idx.values)

with pm.Model() as model:
    sigma_alpha = pm.Uniform('sigma_alpha', lower=0, upper=1, shape=n_country)
    sigma_beta = pm.Uniform('sigma_beta', lower=0, upper=1, shape=n_country)
    
    alpha = mvNormalRandomWalk('alpha',taus=1/sigma_alpha,shape=(time_slots,n_country))
    beta = mvNormalRandomWalk('beta',taus=1/sigma_beta,shape=(time_slots,n_country))
    
    alpha_r = alpha.T[country_idx, time_slot]
    beta_r = beta.T[country_idx, time_slot]

    f = alpha_r + beta_r*t_t
    
    sd = pm.Uniform('sd',lower=0,upper=1,testval=T.std(y_t))
    likelihood = pm.Normal('y',mu=f,sd=sd,observed=y_t)
```
@AustinRochford any comments before I close this issue?

**Comment by AustinRochford** [comment]
Nope, go for it.
